[
  {
    "package": {
      "name": "flask",
      "version": "2.2.2",
      "ecosystem": "PyPI"
    },
    "vulnerabilities": [
      {
        "modified": "2023-08-21T18:29:55Z",
        "published": "2023-05-01T19:22:20Z",
        "schema_version": "1.4.0",
        "id": "GHSA-m2qf-hxjv-5gpq",
        "aliases": [
          "CVE-2023-30861"
        ],
        "summary": "Flask vulnerable to possible disclosure of permanent session cookie due to missing Vary: Cookie header",
        "details": "When all of the following conditions are met, a response containing data intended for one client may be cached and subsequently sent by a proxy to other clients. If the proxy also caches `Set-Cookie` headers, it may send one client's `session` cookie to other clients. The severity depends on the application's use of the session, and the proxy's behavior regarding cookies. The risk depends on _all_ these conditions being met.\n\n1. The application must be hosted behind a caching proxy that does not strip cookies or ignore responses with cookies.\n2. The application sets [`session.permanent = True`](https://flask.palletsprojects.com/en/2.3.x/api/#flask.session.permanent).\n2. The application does not access or modify the session at any point during a request.\n4. [`SESSION_REFRESH_EACH_REQUEST`](https://flask.palletsprojects.com/en/2.3.x/config/#SESSION_REFRESH_EACH_REQUEST) is enabled (the default).\n5. The application does not set a `Cache-Control` header to indicate that a page is private or should not be cached.\n\nThis happens because vulnerable versions of Flask only set the `Vary: Cookie` header when the session is accessed or modified, not when it is refreshed (re-sent to update the expiration) without being accessed or modified.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "flask",
              "purl": "pkg:pypi/flask"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "2.3.0"
                  },
                  {
                    "fixed": "2.3.2"
                  }
                ]
              }
            ],
            "versions": [
              "2.3.0",
              "2.3.1"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/05/GHSA-m2qf-hxjv-5gpq/GHSA-m2qf-hxjv-5gpq.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "flask.sessions.SecureCookieSessionInterface.save_session"
              ]
            }
          },
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "flask",
              "purl": "pkg:pypi/flask"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.5"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.11",
              "0.11.1",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.12.3",
              "0.12.4",
              "0.12.5",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.5",
              "0.5.1",
              "0.5.2",
              "0.6",
              "0.6.1",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.9",
              "1.0",
              "1.0.1",
              "1.0.2",
              "1.0.3",
              "1.0.4",
              "1.1.0",
              "1.1.1",
              "1.1.2",
              "1.1.3",
              "1.1.4",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.2.1",
              "2.2.2",
              "2.2.3",
              "2.2.4"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/05/GHSA-m2qf-hxjv-5gpq/GHSA-m2qf-hxjv-5gpq.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "flask.sessions.SecureCookieSessionInterface.save_session"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N"
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/security/advisories/GHSA-m2qf-hxjv-5gpq"
          },
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-30861"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/commit/70f906c51ce49c485f1d355703e9cc3386b1cc2b"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/commit/afd63b16170b7c047f5758eb910c416511e9c965"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pallets/flask"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/releases/tag/2.2.5"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/releases/tag/2.3.2"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pypa/advisory-database/tree/main/vulns/flask/PYSEC-2023-62.yaml"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2023/08/msg00024.html"
          },
          {
            "type": "WEB",
            "url": "https://security.netapp.com/advisory/ntap-20230818-0006/"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2023/dsa-5442"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-539"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-05-01T19:22:20Z",
          "nvd_published_at": null,
          "severity": "HIGH"
        }
      },
      {
        "modified": "2023-06-05T01:12:52Z",
        "published": "2023-05-02T18:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-62",
        "aliases": [
          "CVE-2023-30861",
          "GHSA-m2qf-hxjv-5gpq"
        ],
        "details": "Flask is a lightweight WSGI web application framework. When all of the following conditions are met, a response containing data intended for one client may be cached and subsequently sent by the proxy to other clients. If the proxy also caches `Set-Cookie` headers, it may send one client's `session` cookie to other clients. The severity depends on the application's use of the session and the proxy's behavior regarding cookies. The risk depends on all these conditions being met.\n\n1. The application must be hosted behind a caching proxy that does not strip cookies or ignore responses with cookies.\n2. The application sets `session.permanent = True`\n3. The application does not access or modify the session at any point during a request.\n4. `SESSION_REFRESH_EACH_REQUEST` enabled (the default).\n5. The application does not set a `Cache-Control` header to indicate that a page is private or should not be cached.\n\nThis happens because vulnerable versions of Flask only set the `Vary: Cookie` header when the session is accessed or modified, not when it is refreshed (re-sent to update the expiration) without being accessed or modified. This issue has been fixed in versions 2.3.2 and 2.2.5.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "flask",
              "purl": "pkg:pypi/flask"
            },
            "ranges": [
              {
                "type": "GIT",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "70f906c51ce49c485f1d355703e9cc3386b1cc2b"
                  },
                  {
                    "fixed": "afd63b16170b7c047f5758eb910c416511e9c965"
                  }
                ],
                "repo": "https://github.com/pallets/flask"
              },
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.5"
                  },
                  {
                    "introduced": "2.3.0"
                  },
                  {
                    "fixed": "2.3.2"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.11",
              "0.11.1",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.12.3",
              "0.12.4",
              "0.12.5",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.5",
              "0.5.1",
              "0.5.2",
              "0.6",
              "0.6.1",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.9",
              "1.0",
              "1.0.1",
              "1.0.2",
              "1.0.3",
              "1.0.4",
              "1.1.0",
              "1.1.1",
              "1.1.2",
              "1.1.3",
              "1.1.4",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.1.3",
              "2.2.0",
              "2.2.1",
              "2.2.2",
              "2.2.3",
              "2.2.4",
              "2.3.0",
              "2.3.1"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/flask/PYSEC-2023-62.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "FIX",
            "url": "https://github.com/pallets/flask/commit/70f906c51ce49c485f1d355703e9cc3386b1cc2b"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/releases/tag/2.3.2"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/flask/releases/tag/2.2.5"
          },
          {
            "type": "ADVISORY",
            "url": "https://github.com/pallets/flask/security/advisories/GHSA-m2qf-hxjv-5gpq"
          },
          {
            "type": "FIX",
            "url": "https://github.com/pallets/flask/commit/afd63b16170b7c047f5758eb910c416511e9c965"
          }
        ]
      }
    ],
    "groups": [
      {
        "ids": [
          "GHSA-m2qf-hxjv-5gpq",
          "PYSEC-2023-62"
        ]
      }
    ]
  },
  {
    "package": {
      "name": "redis",
      "version": "4.4.0",
      "ecosystem": "PyPI"
    },
    "vulnerabilities": [
      {
        "modified": "2023-06-06T18:18:13Z",
        "published": "2023-03-26T21:30:23Z",
        "schema_version": "1.4.0",
        "id": "GHSA-24wv-mv5m-xv4h",
        "aliases": [
          "CVE-2023-28858"
        ],
        "summary": "redis-py Race Condition vulnerability",
        "details": "redis-py before 4.5.3, as used in ChatGPT and other products, leaves a connection open after canceling an async Redis command at an inopportune time (in the case of a pipeline operation), and can send response data to the client of an unrelated request in an off-by-one manner. The fixed versions for this CVE Record are 4.3.6, 4.4.3, and 4.5.3, but [are believed to be incomplete](https://github.com/redis/redis-py/issues/2665). CVE-2023-28859 has been assigned the issues caused by the incomplete fixes.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.4.0"
                  },
                  {
                    "fixed": "4.4.3"
                  }
                ]
              }
            ],
            "versions": [
              "4.4.0",
              "4.4.1",
              "4.4.2"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/03/GHSA-24wv-mv5m-xv4h/GHSA-24wv-mv5m-xv4h.json"
            }
          },
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.5.0"
                  },
                  {
                    "fixed": "4.5.3"
                  }
                ]
              }
            ],
            "versions": [
              "4.5.0",
              "4.5.1",
              "4.5.2"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/03/GHSA-24wv-mv5m-xv4h/GHSA-24wv-mv5m-xv4h.json"
            }
          },
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.2.0"
                  },
                  {
                    "fixed": "4.3.6"
                  }
                ]
              }
            ],
            "versions": [
              "4.2.0",
              "4.2.1",
              "4.2.2",
              "4.3.0",
              "4.3.1",
              "4.3.2",
              "4.3.3",
              "4.3.4",
              "4.3.5"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/03/GHSA-24wv-mv5m-xv4h/GHSA-24wv-mv5m-xv4h.json"
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:L/I:N/A:N"
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-28858"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/issues/2624"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2641"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/commit/d56baeb683fc1935cfa343fa2eeb0fa9bd955283"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pypa/advisory-database/tree/main/vulns/redis/PYSEC-2023-45.yaml"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/redis/redis-py"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.3.5...v4.3.6"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.4.2...v4.4.3"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.5.2...v4.5.3"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/releases/tag/v4.4.4"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/releases/tag/v4.5.4"
          },
          {
            "type": "WEB",
            "url": "https://openai.com/blog/march-20-chatgpt-outage"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-193"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-03-27T21:31:45Z",
          "nvd_published_at": "2023-03-26T19:15:00Z",
          "severity": "LOW"
        }
      },
      {
        "modified": "2023-07-19T22:49:04Z",
        "published": "2023-03-26T21:30:23Z",
        "schema_version": "1.4.0",
        "id": "GHSA-8fww-64cx-x8p5",
        "aliases": [
          "CVE-2023-28859"
        ],
        "summary": "redis-py Race Condition due to incomplete fix",
        "details": "redis-py through 4.5.3 and 4.4.3 leaves a connection open after canceling an async Redis command at an inopportune time (in the case of a non-pipeline operation), and can send response data to the client of an unrelated request. NOTE: this issue exists because of an incomplete fix for CVE-2023-28858.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.5.0"
                  },
                  {
                    "fixed": "4.5.4"
                  }
                ]
              }
            ],
            "versions": [
              "4.5.0",
              "4.5.1",
              "4.5.2",
              "4.5.3"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/03/GHSA-8fww-64cx-x8p5/GHSA-8fww-64cx-x8p5.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "redis.asyncio.client.Pipeline"
              ]
            }
          },
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.2.0"
                  },
                  {
                    "fixed": "4.4.4"
                  }
                ]
              }
            ],
            "versions": [
              "4.2.0",
              "4.2.1",
              "4.2.2",
              "4.3.0",
              "4.3.1",
              "4.3.2",
              "4.3.3",
              "4.3.4",
              "4.3.5",
              "4.3.6",
              "4.4.0",
              "4.4.0rc1",
              "4.4.0rc2",
              "4.4.0rc3",
              "4.4.0rc4",
              "4.4.1",
              "4.4.2",
              "4.4.3"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/03/GHSA-8fww-64cx-x8p5/GHSA-8fww-64cx-x8p5.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "redis.asyncio.client.Pipeline"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:N/A:N"
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-28859"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/issues/2665"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/1899"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2641"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2666"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2671"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/commit/66a4d6b2a493dd3a20cc299ab5fef3c14baad965"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/commit/b3c89acd0ffe8303649ad8207bc911b1d6a033eb"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pypa/advisory-database/tree/main/vulns/redis/PYSEC-2023-46.yaml"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/redis/redis-py"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/releases/tag/v4.4.4"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/releases/tag/v4.5.4"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-459"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-03-27T21:31:52Z",
          "nvd_published_at": "2023-03-26T19:15:00Z",
          "severity": "MODERATE"
        }
      },
      {
        "modified": "2023-06-05T01:13:00Z",
        "published": "2023-03-26T19:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-45",
        "aliases": [
          "CVE-2023-28858"
        ],
        "details": "redis-py before 4.5.3, as used in ChatGPT and other products, leaves a connection open after canceling an async Redis command at an inopportune time (in the case of a pipeline operation), and can send response data to the client of an unrelated request in an off-by-one manner. The fixed versions for this CVE Record are 4.3.6, 4.4.3, and 4.5.3; however, CVE-2023-28859 is a separate vulnerability.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.2.0"
                  },
                  {
                    "fixed": "4.3.6"
                  },
                  {
                    "introduced": "4.4.0"
                  },
                  {
                    "fixed": "4.4.3"
                  },
                  {
                    "introduced": "4.5.0"
                  },
                  {
                    "fixed": "4.5.3"
                  }
                ]
              }
            ],
            "versions": [
              "4.2.0",
              "4.2.1",
              "4.2.2",
              "4.3.0",
              "4.3.1",
              "4.3.2",
              "4.3.3",
              "4.3.4",
              "4.3.5",
              "4.4.0",
              "4.4.1",
              "4.4.2",
              "4.5.0",
              "4.5.1",
              "4.5.2"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/redis/PYSEC-2023-45.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.3.5...v4.3.6"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2641"
          },
          {
            "type": "ARTICLE",
            "url": "https://openai.com/blog/march-20-chatgpt-outage"
          },
          {
            "type": "REPORT",
            "url": "https://github.com/redis/redis-py/issues/2624"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.4.2...v4.4.3"
          },
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/compare/v4.5.2...v4.5.3"
          }
        ]
      },
      {
        "modified": "2023-06-05T01:13:00Z",
        "published": "2023-03-26T19:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-46",
        "aliases": [
          "CVE-2023-28859"
        ],
        "details": "redis-py through 4.5.3 leaves a connection open after canceling an async Redis command at an inopportune time (in the case of a non-pipeline operation), and can send response data to the client of an unrelated request. NOTE: this issue exists because of an incomplete fix for CVE-2023-28858.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "redis",
              "purl": "pkg:pypi/redis"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "4.2.0"
                  },
                  {
                    "fixed": "4.4.4"
                  },
                  {
                    "introduced": "4.5.0"
                  },
                  {
                    "fixed": "4.5.4"
                  }
                ]
              }
            ],
            "versions": [
              "4.2.0",
              "4.2.1",
              "4.2.2",
              "4.3.0",
              "4.3.1",
              "4.3.2",
              "4.3.3",
              "4.3.4",
              "4.3.5",
              "4.3.6",
              "4.4.0",
              "4.4.0rc1",
              "4.4.0rc2",
              "4.4.0rc3",
              "4.4.0rc4",
              "4.4.1",
              "4.4.2",
              "4.4.3",
              "4.5.0",
              "4.5.1",
              "4.5.2",
              "4.5.3"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/redis/PYSEC-2023-46.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/redis/redis-py/pull/2641"
          },
          {
            "type": "REPORT",
            "url": "https://github.com/redis/redis-py/issues/2665"
          }
        ]
      }
    ],
    "groups": [
      {
        "ids": [
          "GHSA-24wv-mv5m-xv4h",
          "PYSEC-2023-45"
        ]
      },
      {
        "ids": [
          "GHSA-8fww-64cx-x8p5",
          "PYSEC-2023-46"
        ]
      }
    ]
  },
  {
    "package": {
      "name": "requests",
      "version": "2.28.1",
      "ecosystem": "PyPI"
    },
    "vulnerabilities": [
      {
        "modified": "2023-06-20T14:40:07Z",
        "published": "2023-05-22T20:36:32Z",
        "schema_version": "1.4.0",
        "id": "GHSA-j8r2-6x86-q33q",
        "aliases": [
          "CVE-2023-32681"
        ],
        "summary": "Unintended leak of Proxy-Authorization header in requests",
        "details": "### Impact\n\nSince Requests v2.3.0, Requests has been vulnerable to potentially leaking `Proxy-Authorization` headers to destination servers, specifically during redirects to an HTTPS origin. This is a product of how `rebuild_proxies` is used to recompute and [reattach the `Proxy-Authorization` header](https://github.com/psf/requests/blob/f2629e9e3c7ce3c3c8c025bcd8db551101cbc773/requests/sessions.py#L319-L328) to requests when redirected. Note this behavior has _only_ been observed to affect proxied requests when credentials are supplied in the URL user information component (e.g. `https://username:password@proxy:8080`).\n\n**Current vulnerable behavior(s):**\n\n1. HTTP \u2192 HTTPS: **leak**\n2. HTTPS \u2192 HTTP: **no leak**\n3. HTTPS \u2192 HTTPS: **leak**\n4. HTTP \u2192 HTTP: **no leak**\n\nFor HTTP connections sent through the proxy, the proxy will identify the header in the request itself and remove it prior to forwarding to the destination server. However when sent over HTTPS, the `Proxy-Authorization` header must be sent in the CONNECT request as the proxy has no visibility into further tunneled requests. This results in Requests forwarding the header to the destination server unintentionally, allowing a malicious actor to potentially exfiltrate those credentials.\n\nThe reason this currently works for HTTPS connections in Requests is the `Proxy-Authorization` header is also handled by urllib3 with our usage of the ProxyManager in adapters.py with [`proxy_manager_for`](https://github.com/psf/requests/blob/f2629e9e3c7ce3c3c8c025bcd8db551101cbc773/requests/adapters.py#L199-L235). This will compute the required proxy headers in `proxy_headers` and pass them to the Proxy Manager, avoiding attaching them directly to the Request object. This will be our preferred option going forward for default usage.\n\n### Patches\nStarting in Requests v2.31.0, Requests will no longer attach this header to redirects with an HTTPS destination. This should have no negative impacts on the default behavior of the library as the proxy credentials are already properly being handled by urllib3's ProxyManager.\n\nFor users with custom adapters, this _may_ be potentially breaking if you were already working around this behavior. The previous functionality of `rebuild_proxies` doesn't make sense in any case, so we would encourage any users impacted to migrate any handling of Proxy-Authorization directly into their custom adapter.\n\n### Workarounds\nFor users who are not able to update Requests immediately, there is one potential workaround.\n\nYou may disable redirects by setting `allow_redirects` to `False` on all calls through Requests top-level APIs. Note that if you're currently relying on redirect behaviors, you will need to capture the 3xx response codes and ensure a new request is made to the redirect destination.\n```\nimport requests\nr = requests.get('http://github.com/', allow_redirects=False)\n```\n\n### Credits\n\nThis vulnerability was discovered and disclosed by the following individuals.\n\nDennis Brinkrolf, Haxolot (https://haxolot.com/)\nTobias Funke, (tobiasfunke93@gmail.com)",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "requests",
              "purl": "pkg:pypi/requests"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "2.3.0"
                  },
                  {
                    "fixed": "2.31.0"
                  }
                ]
              }
            ],
            "versions": [
              "2.10.0",
              "2.11.0",
              "2.11.1",
              "2.12.0",
              "2.12.1",
              "2.12.2",
              "2.12.3",
              "2.12.4",
              "2.12.5",
              "2.13.0",
              "2.14.0",
              "2.14.1",
              "2.14.2",
              "2.15.0",
              "2.15.1",
              "2.16.0",
              "2.16.1",
              "2.16.2",
              "2.16.3",
              "2.16.4",
              "2.16.5",
              "2.17.0",
              "2.17.1",
              "2.17.2",
              "2.17.3",
              "2.18.0",
              "2.18.1",
              "2.18.2",
              "2.18.3",
              "2.18.4",
              "2.19.0",
              "2.19.1",
              "2.20.0",
              "2.20.1",
              "2.21.0",
              "2.22.0",
              "2.23.0",
              "2.24.0",
              "2.25.0",
              "2.25.1",
              "2.26.0",
              "2.27.0",
              "2.27.1",
              "2.28.0",
              "2.28.1",
              "2.28.2",
              "2.29.0",
              "2.3.0",
              "2.30.0",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.4.3",
              "2.5.0",
              "2.5.1",
              "2.5.2",
              "2.5.3",
              "2.6.0",
              "2.6.1",
              "2.6.2",
              "2.7.0",
              "2.8.0",
              "2.8.1",
              "2.9.0",
              "2.9.1",
              "2.9.2"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/05/GHSA-j8r2-6x86-q33q/GHSA-j8r2-6x86-q33q.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "requests.sessions.SessionRedirectMixin.rebuild_proxies"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:C/C:H/I:N/A:N"
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/psf/requests/security/advisories/GHSA-j8r2-6x86-q33q"
          },
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-32681"
          },
          {
            "type": "WEB",
            "url": "https://github.com/psf/requests/commit/74ea7cf7a6a27a4eeb2ae24e162bcc942a6706d5"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/psf/requests"
          },
          {
            "type": "WEB",
            "url": "https://github.com/psf/requests/releases/tag/v2.31.0"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pypa/advisory-database/tree/main/vulns/requests/PYSEC-2023-74.yaml"
          },
          {
            "type": "WEB",
            "url": "https://lists.debian.org/debian-lts-announce/2023/06/msg00018.html"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AW7HNFGYP44RT3DUDQXG2QT3OEV2PJ7Y/"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/KOYASTZDGQG2BWLSNBPL3TQRL2G7QYNZ/"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-200"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-05-22T20:36:32Z",
          "nvd_published_at": null,
          "severity": "MODERATE"
        }
      },
      {
        "modified": "2023-06-05T01:13:00Z",
        "published": "2023-05-26T18:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-74",
        "aliases": [
          "CVE-2023-32681",
          "GHSA-j8r2-6x86-q33q"
        ],
        "details": "Requests is a HTTP library. Since Requests 2.3.0, Requests has been leaking Proxy-Authorization headers to destination servers when redirected to an HTTPS endpoint. This is a product of how we use `rebuild_proxies` to reattach the `Proxy-Authorization` header to requests. For HTTP connections sent through the tunnel, the proxy will identify the header in the request itself and remove it prior to forwarding to the destination server. However when sent over HTTPS, the `Proxy-Authorization` header must be sent in the CONNECT request as the proxy has no visibility into the tunneled request. This results in Requests forwarding proxy credentials to the destination server unintentionally, allowing a malicious actor to potentially exfiltrate sensitive information. This issue has been patched in version 2.31.0.\n\n",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "requests",
              "purl": "pkg:pypi/requests"
            },
            "ranges": [
              {
                "type": "GIT",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "74ea7cf7a6a27a4eeb2ae24e162bcc942a6706d5"
                  }
                ],
                "repo": "https://github.com/psf/requests"
              },
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "2.3.0"
                  },
                  {
                    "fixed": "2.31.0"
                  }
                ]
              }
            ],
            "versions": [
              "2.10.0",
              "2.11.0",
              "2.11.1",
              "2.12.0",
              "2.12.1",
              "2.12.2",
              "2.12.3",
              "2.12.4",
              "2.12.5",
              "2.13.0",
              "2.14.0",
              "2.14.1",
              "2.14.2",
              "2.15.0",
              "2.15.1",
              "2.16.0",
              "2.16.1",
              "2.16.2",
              "2.16.3",
              "2.16.4",
              "2.16.5",
              "2.17.0",
              "2.17.1",
              "2.17.2",
              "2.17.3",
              "2.18.0",
              "2.18.1",
              "2.18.2",
              "2.18.3",
              "2.18.4",
              "2.19.0",
              "2.19.1",
              "2.20.0",
              "2.20.1",
              "2.21.0",
              "2.22.0",
              "2.23.0",
              "2.24.0",
              "2.25.0",
              "2.25.1",
              "2.26.0",
              "2.27.0",
              "2.27.1",
              "2.28.0",
              "2.28.1",
              "2.28.2",
              "2.29.0",
              "2.3.0",
              "2.30.0",
              "2.4.0",
              "2.4.1",
              "2.4.2",
              "2.4.3",
              "2.5.0",
              "2.5.1",
              "2.5.2",
              "2.5.3",
              "2.6.0",
              "2.6.1",
              "2.6.2",
              "2.7.0",
              "2.8.0",
              "2.8.1",
              "2.9.0",
              "2.9.1",
              "2.9.2"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/requests/PYSEC-2023-74.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://github.com/psf/requests/security/advisories/GHSA-j8r2-6x86-q33q"
          },
          {
            "type": "WEB",
            "url": "https://github.com/psf/requests/releases/tag/v2.31.0"
          },
          {
            "type": "FIX",
            "url": "https://github.com/psf/requests/commit/74ea7cf7a6a27a4eeb2ae24e162bcc942a6706d5"
          },
          {
            "type": "WEB",
            "url": "https://lists.fedoraproject.org/archives/list/package-announce@lists.fedoraproject.org/message/AW7HNFGYP44RT3DUDQXG2QT3OEV2PJ7Y/"
          }
        ]
      }
    ],
    "groups": [
      {
        "ids": [
          "GHSA-j8r2-6x86-q33q",
          "PYSEC-2023-74"
        ]
      }
    ]
  },
  {
    "package": {
      "name": "werkzeug",
      "version": "2.2.2",
      "ecosystem": "PyPI"
    },
    "vulnerabilities": [
      {
        "modified": "2023-08-18T16:01:11Z",
        "published": "2023-02-15T15:37:03Z",
        "schema_version": "1.4.0",
        "id": "GHSA-px8h-6qxv-m22q",
        "aliases": [
          "CVE-2023-23934"
        ],
        "summary": "Incorrect parsing of nameless cookies leads to __Host- cookies bypass",
        "details": "Browsers may allow \"nameless\" cookies that look like `=value` instead of `key=value`. A vulnerable browser may allow a compromised application on an adjacent subdomain to exploit this to set a cookie like `=__Host-test=bad` for another subdomain.\n\nWerkzeug <= 2.2.2 will parse the cookie `=__Host-test=bad` as `__Host-test=bad`. If a Werkzeug application is running next to a vulnerable or malicious subdomain which sets such a cookie using a vulnerable browser, the Werkzeug application will see the bad cookie value but the valid cookie key.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "werkzeug",
              "purl": "pkg:pypi/werkzeug"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.3"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.10.2",
              "0.10.3",
              "0.10.4",
              "0.11",
              "0.11.1",
              "0.11.10",
              "0.11.11",
              "0.11.12",
              "0.11.13",
              "0.11.14",
              "0.11.15",
              "0.11.2",
              "0.11.3",
              "0.11.4",
              "0.11.5",
              "0.11.6",
              "0.11.7",
              "0.11.8",
              "0.11.9",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.13",
              "0.14",
              "0.14.1",
              "0.15.0",
              "0.15.1",
              "0.15.2",
              "0.15.3",
              "0.15.4",
              "0.15.5",
              "0.15.6",
              "0.16.0",
              "0.16.1",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.4.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.6.1",
              "0.6.2",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.8.2",
              "0.8.3",
              "0.9",
              "0.9.1",
              "0.9.2",
              "0.9.3",
              "0.9.4",
              "0.9.5",
              "0.9.6",
              "1.0.0",
              "1.0.0rc1",
              "1.0.1",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.0rc3",
              "2.0.0rc4",
              "2.0.0rc5",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.2.0",
              "2.2.0a1",
              "2.2.1",
              "2.2.2"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/02/GHSA-px8h-6qxv-m22q/GHSA-px8h-6qxv-m22q.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "werkzeug.http.parse_cookie"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:A/AC:H/PR:N/UI:R/S:U/C:N/I:L/A:N"
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/security/advisories/GHSA-px8h-6qxv-m22q"
          },
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-23934"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/commit/cf275f42acad1b5950c50ffe8ef58fe62cdce028"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pallets/werkzeug"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/releases/tag/2.2.3"
          },
          {
            "type": "WEB",
            "url": "https://security.netapp.com/advisory/ntap-20230818-0003/"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2023/dsa-5470"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-20"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-02-15T15:37:03Z",
          "nvd_published_at": "2023-02-14T20:15:00Z",
          "severity": "LOW"
        }
      },
      {
        "modified": "2023-08-18T16:01:28Z",
        "published": "2023-02-15T15:36:26Z",
        "schema_version": "1.4.0",
        "id": "GHSA-xg9f-g7g7-2323",
        "aliases": [
          "CVE-2023-25577"
        ],
        "summary": "High resource usage when parsing multipart form data with many fields",
        "details": "Werkzeug's multipart form data parser will parse an unlimited number of parts, including file parts. Parts can be a small amount of bytes, but each requires CPU time to parse and may use more memory as Python data. If a request can be made to an endpoint that accesses `request.data`, `request.form`, `request.files`, or `request.get_data(parse_form_data=False)`, it can cause unexpectedly high resource usage.\n\nThis allows an attacker to cause a denial of service by sending crafted multipart data to an endpoint that will parse it. The amount of CPU time required can block worker processes from handling legitimate requests. The amount of RAM required can trigger an out of memory kill of the process. Unlimited file parts can use up memory and file handles. If many concurrent requests are sent continuously, this can exhaust or kill all available workers.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "werkzeug",
              "purl": "pkg:pypi/werkzeug"
            },
            "ranges": [
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.3"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.10.2",
              "0.10.3",
              "0.10.4",
              "0.11",
              "0.11.1",
              "0.11.10",
              "0.11.11",
              "0.11.12",
              "0.11.13",
              "0.11.14",
              "0.11.15",
              "0.11.2",
              "0.11.3",
              "0.11.4",
              "0.11.5",
              "0.11.6",
              "0.11.7",
              "0.11.8",
              "0.11.9",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.13",
              "0.14",
              "0.14.1",
              "0.15.0",
              "0.15.1",
              "0.15.2",
              "0.15.3",
              "0.15.4",
              "0.15.5",
              "0.15.6",
              "0.16.0",
              "0.16.1",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.4.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.6.1",
              "0.6.2",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.8.2",
              "0.8.3",
              "0.9",
              "0.9.1",
              "0.9.2",
              "0.9.3",
              "0.9.4",
              "0.9.5",
              "0.9.6",
              "1.0.0",
              "1.0.0rc1",
              "1.0.1",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.0rc3",
              "2.0.0rc4",
              "2.0.0rc5",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.2.0",
              "2.2.0a1",
              "2.2.1",
              "2.2.2"
            ],
            "database_specific": {
              "source": "https://github.com/github/advisory-database/blob/main/advisories/github-reviewed/2023/02/GHSA-xg9f-g7g7-2323/GHSA-xg9f-g7g7-2323.json"
            },
            "ecosystem_specific": {
              "affected_functions": [
                "werkzeug.Request.from_values"
              ]
            }
          }
        ],
        "severity": [
          {
            "type": "CVSS_V3",
            "score": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
          }
        ],
        "references": [
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/security/advisories/GHSA-xg9f-g7g7-2323"
          },
          {
            "type": "ADVISORY",
            "url": "https://nvd.nist.gov/vuln/detail/CVE-2023-25577"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/commit/517cac5a804e8c4dc4ed038bb20dacd038e7a9f1"
          },
          {
            "type": "PACKAGE",
            "url": "https://github.com/pallets/werkzeug"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/releases/tag/2.2.3"
          },
          {
            "type": "WEB",
            "url": "https://security.netapp.com/advisory/ntap-20230818-0003/"
          },
          {
            "type": "WEB",
            "url": "https://www.debian.org/security/2023/dsa-5470"
          }
        ],
        "database_specific": {
          "cwe_ids": [
            "CWE-400",
            "CWE-770"
          ],
          "github_reviewed": true,
          "github_reviewed_at": "2023-02-15T15:36:26Z",
          "nvd_published_at": "2023-02-14T20:15:00Z",
          "severity": "HIGH"
        }
      },
      {
        "modified": "2023-05-04T03:49:48Z",
        "published": "2023-02-14T20:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-57",
        "aliases": [
          "CVE-2023-23934",
          "GHSA-px8h-6qxv-m22q"
        ],
        "details": "Werkzeug is a comprehensive WSGI web application library. Browsers may allow \"nameless\" cookies that look like `=value` instead of `key=value`. A vulnerable browser may allow a compromised application on an adjacent subdomain to exploit this to set a cookie like `=__Host-test=bad` for another subdomain. Werkzeug prior to 2.2.3 will parse the cookie `=__Host-test=bad` as __Host-test=bad`. If a Werkzeug application is running next to a vulnerable or malicious subdomain which sets such a cookie using a vulnerable browser, the Werkzeug application will see the bad cookie value but the valid cookie key. The issue is fixed in Werkzeug 2.2.3.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "werkzeug",
              "purl": "pkg:pypi/werkzeug"
            },
            "ranges": [
              {
                "type": "GIT",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "cf275f42acad1b5950c50ffe8ef58fe62cdce028"
                  }
                ],
                "repo": "https://github.com/pallets/werkzeug"
              },
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.3"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.10.2",
              "0.10.3",
              "0.10.4",
              "0.11",
              "0.11.1",
              "0.11.10",
              "0.11.11",
              "0.11.12",
              "0.11.13",
              "0.11.14",
              "0.11.15",
              "0.11.2",
              "0.11.3",
              "0.11.4",
              "0.11.5",
              "0.11.6",
              "0.11.7",
              "0.11.8",
              "0.11.9",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.13",
              "0.14",
              "0.14.1",
              "0.15.0",
              "0.15.1",
              "0.15.2",
              "0.15.3",
              "0.15.4",
              "0.15.5",
              "0.15.6",
              "0.16.0",
              "0.16.1",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.4.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.6.1",
              "0.6.2",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.8.2",
              "0.8.3",
              "0.9",
              "0.9.1",
              "0.9.2",
              "0.9.3",
              "0.9.4",
              "0.9.5",
              "0.9.6",
              "1.0.0",
              "1.0.0rc1",
              "1.0.1",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.0rc3",
              "2.0.0rc4",
              "2.0.0rc5",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.2.0",
              "2.2.0a1",
              "2.2.1",
              "2.2.2"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/werkzeug/PYSEC-2023-57.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "ADVISORY",
            "url": "https://github.com/pallets/werkzeug/security/advisories/GHSA-px8h-6qxv-m22q"
          },
          {
            "type": "FIX",
            "url": "https://github.com/pallets/werkzeug/commit/cf275f42acad1b5950c50ffe8ef58fe62cdce028"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/releases/tag/2.2.3"
          }
        ]
      },
      {
        "modified": "2023-05-04T03:49:48Z",
        "published": "2023-02-14T20:15:00Z",
        "schema_version": "1.4.0",
        "id": "PYSEC-2023-58",
        "aliases": [
          "CVE-2023-25577",
          "GHSA-xg9f-g7g7-2323"
        ],
        "details": "Werkzeug is a comprehensive WSGI web application library. Prior to version 2.2.3, Werkzeug's multipart form data parser will parse an unlimited number of parts, including file parts. Parts can be a small amount of bytes, but each requires CPU time to parse and may use more memory as Python data. If a request can be made to an endpoint that accesses `request.data`, `request.form`, `request.files`, or `request.get_data(parse_form_data=False)`, it can cause unexpectedly high resource usage. This allows an attacker to cause a denial of service by sending crafted multipart data to an endpoint that will parse it. The amount of CPU time required can block worker processes from handling legitimate requests. The amount of RAM required can trigger an out of memory kill of the process. Unlimited file parts can use up memory and file handles. If many concurrent requests are sent continuously, this can exhaust or kill all available workers. Version 2.2.3 contains a patch for this issue.",
        "affected": [
          {
            "package": {
              "ecosystem": "PyPI",
              "name": "werkzeug",
              "purl": "pkg:pypi/werkzeug"
            },
            "ranges": [
              {
                "type": "GIT",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "517cac5a804e8c4dc4ed038bb20dacd038e7a9f1"
                  }
                ],
                "repo": "https://github.com/pallets/werkzeug"
              },
              {
                "type": "ECOSYSTEM",
                "events": [
                  {
                    "introduced": "0"
                  },
                  {
                    "fixed": "2.2.3"
                  }
                ]
              }
            ],
            "versions": [
              "0.1",
              "0.10",
              "0.10.1",
              "0.10.2",
              "0.10.3",
              "0.10.4",
              "0.11",
              "0.11.1",
              "0.11.10",
              "0.11.11",
              "0.11.12",
              "0.11.13",
              "0.11.14",
              "0.11.15",
              "0.11.2",
              "0.11.3",
              "0.11.4",
              "0.11.5",
              "0.11.6",
              "0.11.7",
              "0.11.8",
              "0.11.9",
              "0.12",
              "0.12.1",
              "0.12.2",
              "0.13",
              "0.14",
              "0.14.1",
              "0.15.0",
              "0.15.1",
              "0.15.2",
              "0.15.3",
              "0.15.4",
              "0.15.5",
              "0.15.6",
              "0.16.0",
              "0.16.1",
              "0.2",
              "0.3",
              "0.3.1",
              "0.4",
              "0.4.1",
              "0.5",
              "0.5.1",
              "0.6",
              "0.6.1",
              "0.6.2",
              "0.7",
              "0.7.1",
              "0.7.2",
              "0.8",
              "0.8.1",
              "0.8.2",
              "0.8.3",
              "0.9",
              "0.9.1",
              "0.9.2",
              "0.9.3",
              "0.9.4",
              "0.9.5",
              "0.9.6",
              "1.0.0",
              "1.0.0rc1",
              "1.0.1",
              "2.0.0",
              "2.0.0rc1",
              "2.0.0rc2",
              "2.0.0rc3",
              "2.0.0rc4",
              "2.0.0rc5",
              "2.0.1",
              "2.0.2",
              "2.0.3",
              "2.1.0",
              "2.1.1",
              "2.1.2",
              "2.2.0",
              "2.2.0a1",
              "2.2.1",
              "2.2.2"
            ],
            "database_specific": {
              "source": "https://github.com/pypa/advisory-database/blob/main/vulns/werkzeug/PYSEC-2023-58.yaml"
            }
          }
        ],
        "references": [
          {
            "type": "FIX",
            "url": "https://github.com/pallets/werkzeug/commit/517cac5a804e8c4dc4ed038bb20dacd038e7a9f1"
          },
          {
            "type": "WEB",
            "url": "https://github.com/pallets/werkzeug/releases/tag/2.2.3"
          },
          {
            "type": "ADVISORY",
            "url": "https://github.com/pallets/werkzeug/security/advisories/GHSA-xg9f-g7g7-2323"
          }
        ]
      }
    ],
    "groups": [
      {
        "ids": [
          "GHSA-px8h-6qxv-m22q",
          "PYSEC-2023-57"
        ]
      },
      {
        "ids": [
          "GHSA-xg9f-g7g7-2323",
          "PYSEC-2023-58"
        ]
      }
    ]
  }
]